### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] enum false positive(s) in 8.59.0

### ISSUE THREAD

**Issue by joshkel** [typescript-eslint/typescript-eslint#12271]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.2.2&fileType=.tsx&code=KYOwrgtgBAwg9gGzgJygbygJWAEwDRQDiywoBAQgmMFAL4BQAxnCAM4AuUAbgIZXCsAXFmDNkOADzwkyAh2QBLEAHMAfFAC86elCgBtaSgB02HAF1hAcgDEAMwAM9y3h37DyI8VIgLUG-YdnVwNEY0pqX38AoIZ6WzAQRnYFFihgAA8eCAAHBGAARgAKXn4AZXZkYXklZQBKbV0AekaoABUACwVWKAB3FABrXCglKAAOIwBWcfSjKABJKGYwBBwoHmRlaih2dp5OcYmATktukGAe3vWQGtdm4e6xEiThHZoEJRouqFsPqAAjACew3YrGACFsBB4IFWrxIJzWUFYChyeVuLT6yH66zgCRwRlczDYnGYMk0UAA1CVqOVUDxuu4ANyuKkCPQklBmMmWMDZHB7XCWJmxeKJZKpYA8dIo4AAJmKfGpFSqFRq9TQaLa7RoGKxyBx0O%2BKG2Wqg2UURoyWVywAI7B4-RqaxwvBAduUNDgtg1rQB2WApUYimy7HhoEgUCQcFBZ1YrHxukJHEWoUqsBTZMpCv9FSZuhZrDZKc5Wm5vP5OEF9GFCSSKRAUB5fPYwAAzPKykrESqVGrvZ1uqDgBBuu9Bgj3eBfrY%2BKDTVGFMkuJ96wdDvHuFmC5mO7T6UWuY3y5XaEA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eFYDAruuGAL4g9A&tokens=false

### Repro Code

```TypeScript
enum Color { Red, Green, Blue }
const values: Record<Color, string> = {
  [Color.Red]: '#f00',
  [Color.Green]: '#0f0',
  [Color.Blue]: '#00f',
}

function example1(valueStr: string) {
  // This worked in 8.58.x, fails in 8.59. Maybe okay: the line is fine by
  // itself, due to TypeScript enum looseness, and there's a simple workaround.
  const color = +valueStr as Color;
  values[color] = 'updated';
}

function eaxmple2(valueStr: string) {
  // Workaround for the prior example
  const color: Color = +valueStr;
  values[color] = 'updated';
}

function update3(valueStr: string) {
  // This seems like a genuine false positive in 8.59: assertion is
  // necessary to prevent TypeScript error.
  values[+valueStr as Color] = 'updated';
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strict": true
  }
}
```

### Expected Result

No typescript-eslint warning in `example3`. `example1` is debatable.

### Actual Result

"This assertion is unnecessary since it does not change the type of the expression" for `example1` and `example3`.

**Comment by labeebahmad201** [comment]
Fix PR: https://github.com/typescript-eslint/typescript-eslint/pull/12272

This addresses the false positives reported in this issue for number-to-enum assertions in no-unnecessary-type-assertion, and adds regression tests for the provided repro cases.

**Comment by joshkel** [comment]
I mentioned that this variant is rejected by typescript-eslint 8.59.0 and that may be okay:

```ts
  const color = +valueStr as Color;
```

After thinking about it further, I think it's best if typescript-eslint accepts it, because it accepts "narrowing" assertions for other types:

```ts
type ValuePath = 'values' | `values.${string}`;
function update(ids: string[]) {
  const value = `values.${ids[0]}` as ValuePath;
  // ...
}
```

**Comment by terencehonles** [comment]
I added this to #12272, but putting it here for additional visibility

I believe this is related, but I'm seeing some conflicts with `@typescript-eslint/no-unsafe-enum-comparison` and I'm wondering if this rule can leverage the other's logic to allow asserting that the enum is of the same type when it's not a literal/const value.

Example:

```TypeScript
enum Color { Red, Green, Blue }

function isRed(value: number) {
  return (value as Color) === Color.Red;
}
```

**Comment by repulsio** [comment]
@bradzacher Thank you for helping to look into similar issues.

I was just confused about the Label "accepting prs". There clearly was a regression starting in 8.59.0. Before this version, my project had no ESLint issues. After this version, my project has countless incorrect `no-unnecessary-type-assertion` issues, and I had to downgrade and hard pin to 8.58.2 for my project.

Instead of "accepting prs", shouldn't this be **"revert"** the commit(s) that started causing this?

Rather than continue building on top of a shaky foundation, I believe we should correct the underlying problem before moving on.

**Comment by bradzacher** [comment]
@repulsio no. See https://github.com/typescript-eslint/typescript-eslint/issues/12250#issuecomment-4289343557

**Comment by repulsio** [comment]
I see this was the strategy:

https://github.com/typescript-eslint/typescript-eslint/pull/11789#issuecomment-3761041270
Expect things to break and fix false positives one by one (in hopes of forcing correctness faster)

Understandable, but I'll have to stay hard-pinned to 8.58 and hope that these false positives are ironed out by the next time I attempt upgrade.
